ca/provisioner.go

```diff
@@ -0,0 +1,194 @@
+	"github.com/smallstep/cli/crypto/randutil"
+	"github.com/smallstep/cli/jose"
+	"github.com/smallstep/cli/token"
+	"github.com/smallstep/cli/token/provision"
+)
+
+const (
+	tokenLifetime = 5 * time.Minute
+)
+
+// Provisioner is an authorized entity that can sign tokens necessary for
+// signature requests.
+type Provisioner struct {
+	name          string
+	kid           string
+	caURL         string
+	caRoot        string
+	jwk           *jose.JSONWebKey
+	tokenLifetime time.Duration
+}
+
+// NewProvisioner loads and decrypts key material from the CA for the named
+// provisioner. The key identified by `kid` will be used if specified. If `kid`
+// is the empty string we'll use the first key for the named provisioner that
+// decrypts using `passFile`.
+func NewProvisioner(name, kid, caURL, caRoot string, password []byte) (*Provisioner, error) {
+	var jwk *jose.JSONWebKey
+	var err error
+	if kid != "" {
+		jwk, err = loadProvisionerJWKByKid(kid, caURL, caRoot, password)
+	} else {
+		jwk, err = loadProvisionerJWKByName(name, caURL, caRoot, password)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return &Provisioner{
+		name:          name,
+		kid:           jwk.KeyID,
+		caURL:         caURL,
+		caRoot:        caRoot,
+		jwk:           jwk,
+		tokenLifetime: tokenLifetime,
+	}, nil
+}
+
+// Name returns the provisioner's name.
+func (p *Provisioner) Name() string {
+	return p.name
+}
+
+// Kid returns the provisioners key ID.
+func (p *Provisioner) Kid() string {
+	return p.kid
+}
+
+// Token generates a bootstrap token for a subject.
+func (p *Provisioner) Token(subject string) (string, error) {
+	// A random jwt id will be used to identify duplicated tokens
+	jwtID, err := randutil.Hex(64) // 256 bits
+	if err != nil {
+		return "", err
+	}
+
+	notBefore := time.Now()
+	notAfter := notBefore.Add(tokenLifetime)
+	signURL := fmt.Sprintf("%v/1.0/sign", p.caURL)
+
+	tokOptions := []token.Options{
+		token.WithJWTID(jwtID),
+		token.WithKid(p.kid),
+		token.WithIssuer(p.name),
+		token.WithAudience(signURL),
+		token.WithValidity(notBefore, notAfter),
+		token.WithRootCA(p.caRoot),
+		token.WithSANS([]string{subject}),
+	}
+
+	tok, err := provision.New(subject, tokOptions...)
+	if err != nil {
+		return "", err
+	}
+
+	return tok.SignedString(p.jwk.Algorithm, p.jwk.Key)
+}
+
+func decryptProvisionerJWK(encryptedKey string, password []byte) (*jose.JSONWebKey, error) {
+	enc, err := jose.ParseEncrypted(encryptedKey)
+	if err != nil {
+		return nil, err
+	}
+	data, err := enc.Decrypt(password)
+	if err != nil {
+		return nil, err
+	}
+	jwk := new(jose.JSONWebKey)
+	if err := json.Unmarshal(data, jwk); err != nil {
+		return nil, errors.Wrap(err, "error unmarshaling provisioning key")
+	}
+	return jwk, nil
+}
+
+// loadProvisionerJWKByKid retrieves a provisioner key from the CA by key ID and
+// decrypts it using the specified password file.
+func loadProvisionerJWKByKid(kid, caURL, caRoot string, password []byte) (*jose.JSONWebKey, error) {
```

Comment:
just a note: a kid does not uniquely identify a provisioner. The password will still probably decrypt it, but it may not be the provisioner that the user intended (e.g. could have different claims).

Reply:
In this case all you're getting is the JWK so it seems like the provisioner claims don't really matter?

Reply:
I got the autocert current behavior, but I will change it to use the name if it's not blank.

Reply:
if fact, no, this method does a `GET /provisioners/<kid>/encrypted-key` that returns only one encrypted key, even if multiple provisioners are defined. Authority assumes they are the same key, here the name is not required.
But it will be required to build the token. So, we need to make sure the name is always passed.